mlpp_lib/model_selection.py

```diff
@@ -1,10 +1,329 @@
+        - "sequential": splits the `time_index` sequentially based on
+            the partition fractions
+
+    station_split_method: string, optional
+        The method to split along time. Currently implemented are:
+        - "random": splits the `station_index` randomly based on
+            the partition fractions
+
+
+    Examples
+    --------
+    >>> splitter = DataSplitter(
+    ...     time_split = {"train": 0.6, "val": 0.2, "test": 0.2},
+    ...     time_split_method = "sequential"
+    ... )
+    ... train_features, train_targets = splitter.get_partition(
+    ...     features, targets, partition="train"
+    ... )
+    """
+
+    partitions: dict[str, dict[str, xindex]]
+    time_split_methods = ["sequential"]
+    station_split_methods = ["random", "sequential"]
+
+    def __init__(
+        self,
+        time_split: dict[str, Any],
+        station_split: Optional[dict[str, Any]],
+        time_split_method: Optional[str] = None,
+        station_split_method: Optional[str] = None,
+    ):
+
+        if not time_split.keys() == station_split.keys():
+            raise ValueError(
+                "Time split and station split must be defined "
+                "with the same partitions!"
+            )
+        self.partition_names = list(time_split.keys())
+        self._check_time(time_split, time_split_method)
+        self._check_station(station_split, station_split_method)
+
+
+    @classmethod
+    def from_json(cls, file: str) -> Self:
+        """Instantiate the DataSplitter from a json file with previously computed splits."""
+
+        with open(file, "r") as f:
+            splits = json.load(f)
+
+        time_split = {k: v["forecast_reference_time"] for k, v in splits.items()}
+        station_split = {k: v["station"] for k, v in splits.items()}
+        splitter = cls(time_split, station_split)
+        splitter._time_defined = True
+        splitter._station_defined = True
+        splitter._time_partitioning()
+        splitter._station_partitioning()
+        return splitter
+
+    def to_json(self, file: str):
+        if not hasattr(self, "partitions"):
+            self._time_partitioning()
+            self._station_partitioning()
+        with open(file, "w") as f:
+            json.dump(self.partitions, f, indent=4)
+
+    def get_partition(
+        self, *args: xr.Dataset, partition=None, thinning: Optional[Mapping] = None
+    ) -> tuple[xr.Dataset, ...]:
+        """
+        Select and return a partition based on the current values in
+        the `partitions` attribute.
+
+        Parameters
+        ----------
+        *args: `xr.Dataset`
+            The dataset arguments from which we want to select a partition.
+        partition: str
+            The keyword for the partition, e.g. "train", "val" or "test".
+
+        Returns
+        -------
+        partition: tuple of `xr.Dataset`
+            Subsets of the input datasets.
+        """
+
+        if partition is None:
```

Comment:
It's optional because I have the *args argument before it, which allows me to pass any number of datasets, but also forces me to make partition "optional", even though it's not. It's undesirable but I believe it's a minor inconvenience as long as the exception is raised. But feel free to suggest another alternative! I'm also not satisfied with the current solution...